### PR TITLE
fix(auth): accept equivalencies when comparing scopes

### LINKED ISSUE #323: Improvement Request: update some endpoints to newest scope
Last February, 3 API endpoints were updated to use the `channel:manage:broadcast` scope. Even though the old `user:edit:broadcast` is still valid, it may be a good idea to add the new scope too, as it's indeed usable and the official Twitch docs refer exclusively to it (and so it may be confusing for some users expecting it to work)

For reference, extracted from [Twitch Changelog](https://dev.twitch.tv/docs/change-log) (2021-02-26):

> The following endpoints were updated to use the channel:manage:broadcast scope. The previous scope remains valid.
> 
> Modify Channel Information
> Create Stream Marker
> Replace Stream Tags

### LINKED ISSUE #342: Pubsub channel subscription scope has changed
## Bug Report

We (Twitch) have changed the pubsub scope for `channel-subscribe-events-v1.<channel ID>` from `channel_subscriptions` to `channel:read:subscriptions`. Twurple fails on a token with this scope. The current scope requirements for topics can be found here: https://dev.twitch.tv/docs/pubsub#topics

### Code

literally any code implementing pubsub for subscriptions

### Expected behavior

pubsub connection works

### Actual Behavior

The following output is from [nodecg-io-twitch-pubsub](https://github.com/codeoverflow-org/nodecg-io/blob/main/services/nodecg-io-twitch-pubsub/package.json):

```
[2022-03-26T00:27:30.565Z] ❌ twurple:pubsub ERROR Retrieving an access token failed: This token does not have the requested scopes (channel_subscriptions) and can not be upgraded.
If you need dynamically upgrading scopes, please implement the AuthProvider interface accordingly:

        https://twurple.js.org/reference/auth/interfaces/AuthProvider.html
[2022-03-26T00:27:30.566Z] ⚠️  twurple:pubsub WARNING No valid token available; trying to refresh
```

### Environment

- Version: 5.0.18
- Node version: v16.14.2
- Operating system: All

### PR DESCRIPTION
**Pull request by iProdigy**
Type: Bugfix

Fixes #342
Fixes #323
Closes #335

* Changes `compareScopes` from an `O(n*m)` operation to `O(n+m)`
* Accept kraken/helix scope equivalencies for `compareScopes`
* Accept `user:edit:broadcast` for `channel:manage:broadcast` in `compareScopes`
* `channel_subscriptions` => `channel:read:subscriptions` for `channel-subscribe-events-v1`
* `user:edit:broadcast` => `channel:manage:broadcast` for `updateChannelInfo`, `createStreamMarker`, and `replaceStreamTags`

### CODE REVIEW

**Comment by theca11** [comment]
Could this be used to fix #323 too? (user:edit:broadcast -> channel:manage:broadcast)

**Comment by iProdigy** [comment]
> Could this be used to fix #323 too? (user:edit:broadcast -> channel:manage:broadcast)

Yes, but extra care would be needed to not accept the reverse (for example in helix `Update User Extensions`) -- that said, the library doesn't appear to currently implement this endpoint so it may be a non-issue

File: packages/auth/src/helpers.ts [hunk 1]
```diff
@@ -190,20 +190,45 @@ export async function getValidTokenFromProvider(
...
+	['user:edit:broadcast', 'channel:manage:broadcast'],
+]);
+
 /**
  * Compares scopes for a non-upgradable `AuthProvider` instance.
  *
  * @param scopesToCompare The scopes to compare against.
  * @param requestedScopes The scopes you requested.
  */
 export function compareScopes(scopesToCompare: string[], requestedScopes?: string[]): void {
```
Comment (iProdigy) [inline packages/auth/src/helpers.ts:212]: Note: in the future, if there is ever an endpoint that accepts `channel:manage:broadcast` but not `user:edit:broadcast`, `strict: boolean = false` could be added to method definitions to skip checking equivalencies when true

**Comment by d-fischer** [comment]
> Could this be used to fix #323 too? (user:edit:broadcast -> channel:manage:broadcast)

Yes, this fixes #323 and supersedes #325 now. I think this is much more elegant and accomplishes the same without ever requiring a breaking change.

File: packages/auth/src/helpers.ts (hunk 1 above)
Comment (d-fischer) [reply packages/auth/src/helpers.ts:212]: Not a pressing issue right now. From past experience with the Twitch API, if they add something new with at least half a year difference to the last thing in a generally similar direction, they'll just add a new scope.

**Comment by d-fischer** [comment]
> Edit: depending on what scope validation the chat package does, the reverse mappings could be eliminated
> 
> Edit2: looks like there's already a `legacyScopes` flag for the chat package, so a helix => legacy mapping doesn't seem to be necessary

Indeed, chat has to do a very strict validation here, since the required scope depends entirely on when the application was created on the Twitch dev portal. From my tests back when the change was introduced, "new" apps don't support the "old" scopes __and vice versa__.

**Pull request merged**